### DISCUSSION TITLE
reference/value in for loop

### DISCUSSION THREAD

**Discussion by xinhaikang** [carbon-language/carbon-lang#1884] [Language design]
Given the below sample code, it seems `e` is a reference to the element in `s`, this will have performance penalty if e is a simple primitive such as integer. in C++, we can specify if we want to iterate with reference or value.

for (e : T in s) {
    if (e <= s.Last()) {
        ++i;
        Swap(&s[i], &e)
    }
}

**Comment by jonmeow** [comment]
This is a difference between C++ and Carbon. You're right that you need to specify in C++; bear in mind that this doesn't work well in templates though, where one or the other needs to be chosen. Developers also aren't universally aware of the performance tradeoff of a given choice. The thought is that maybe we can make it so Carbon better at making the right choice without requiring the developer to fully understand and correctly specify their choice, also making generics and templates perform better in general.

**Comment by josh11b** [comment]
The specifics of whether binding are copies, moves, or references is discussed in #821 (still pending) , #1830 , and https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/README.md#value-categories-and-value-phases (provisional)

**Reply by tkoeppe** [reply]
Regarding

> R-values may not have dedicated storage. This means they cannot be modified and their address generally cannot be taken.

Can one call member functions of r-values, like `MakeThing().Foo();`? The member funciton would presumably expect some form of storage (for the class members), and allow mutation, too -- are we distinguishing const from mutable r-values?

**Reply by josh11b** [reply]
One can call member functions on r-values as long as the function is defined as taking its `me` parameter by value, as in `fn Foo[me: Self]()`. Methods that are declared that way do not allow mutation of `me`. R-values can have members even without dedicated storage. For example, it would be legal for the members to be passed in separate registers.
